examples/tc_port_whitelist/src/bpf/tc.bpf.c

```diff
@@ -0,0 +1,256 @@
+    if (!iph || !data_end) {
+        return NULL;
+    }
+
+    if ((void*)(iph + hdr_sz + sizeof(*tcph)) > data_end) {
+        return NULL;
+    }
+
+    int proto = -1;
+    if (hdr_sz == sizeof(struct iphdr)) {
+        struct iphdr *v4 = (struct iphdr*)iph;
+        proto = v4->protocol;
+    } else if (hdr_sz == sizeof(struct ipv6hdr)) {
+        struct ipv6hdr *v6 = (struct ipv6hdr*)iph;
+        proto = v6->nexthdr;
+    }
+
+    if (proto == IPPROTO_TCP) {
+        tcph = (struct tcphdr*)((void*)iph + hdr_sz);
+    }
+    return tcph;
+}
+
+/*
+ * Allow a port if it is on the allow_ports map
+ *
+ * @param port, value of incoming (ingress case) or outgoing (egress case)
+ *        port in network byte order
+ *
+ * @return true if port is on the allow_ports list
+ *         false otherwise
+ */
+static bool allow_port(__be16 port)
+{
+    u16 hport = bpf_ntohs(port);
+    u32 i = 0;
+    for (i = 0; i < 10; i++) {
+        u32 key = i;
+        u16 *allow_port = bpf_map_lookup_elem(&ports, &key);
+        if (allow_port && hport == *allow_port) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+/*
+ * handle TC hook program
+ *
+ * @param - skb - the bpf socket buffer mirror
+ * @param - rc_allow - Action to take if this packet is on the allow_ports list
+ *          (see TC_ACT_* values in pkt_cls.h)
+ * @param - rc_disallow - Action to take if this packet is not on the allow_ports list
+ *          (see TC_ACT_* values in pkt_cls.h)
+ * 
+ * @return - returns value from rc_disallow if this packet is tcp/ip or udp/ip
+ *           and not on an allowed port
+ *           returns value from rc_allow if this is a tcp/ip udp/ip packet and port
+ *           is on the allow_ports list
+ *           returns TC_ACT_UNSPEC (keep processing packet in TC chain) if this
+ *           packet is not tcp/ip or udp/ip
+ */
+SEC("tc")
+int handle_tc(struct __sk_buff *skb)
+{
+    // default drop packets
+    int rc = rc_disallow; 
+
+    void *data_end = (void*)(long)skb->data_end;
+    struct ethhdr *eth = (struct ethhdr*)(void*)(long)skb->data;
+    struct iphdr *iph = is_ipv4(eth, data_end);
+    struct ipv6hdr *iph6 = is_ipv6(eth, data_end);
+
+    struct udphdr *udph = NULL;
+    struct tcphdr *tcph = NULL;
+    __be16 dst = 0;
+    __be16 src = 0;
+    __be16 port = 0;
+    // IPv4 packet
+    if (iph) {
+        u8 hdr_sz = sizeof(*iph);
+        udph = is_udp(iph, hdr_sz, data_end);
+        tcph = is_tcp(iph, hdr_sz, data_end);
+    } else if (iph6) { // IPv6 packet
+        u8 hdr_sz = sizeof(*iph6);
+        udph = is_udp(iph6, hdr_sz, data_end);
+        tcph = is_tcp(iph6, hdr_sz, data_end);
+    }
+
+    // if both NULL then this was not IPvX/TCP or UDP -- allow
+    if (!udph && !tcph) {
+        goto allow_unknown;
+    }
+
+    if (tcph) {
+        dst = tcph->dest;
+        src = tcph->source;
+    } else if (udph) {
+        dst = udph->dest;
+        src = udph->source;
+    }
+
```

Comment:
I think we can greatly simplify this file if we just do the parsing in this function instead of defining `is_ipv4`, `is_ipv6`, `is_udp` and `is_tcp` in separate functions. I think we can just do this:

```
  __u8 proto = 0;
  void *trans_data;
  if (eth + 1 > data_end) 
      return TC_ACT_UNSPEC;
  if (eth->h_proto == bpf_htons(ETH_P_IP)) { // ipv4
      struct iphdr *iph = (struct iphdr *)(eth + sizeof(*eth));
      if (iph + 1 > data_end)
          return TC_ACT_UNSPEC;
      proto = iph->protocol;
      trans_data = iph + (iph->ihl * 4);
  } else if (eth->h_proto == bpf_htons(ETH_P_IPV6)) { // ipv6
      struct ipv6hdr *ip6h = (struct ipv6hdr *)(data + sizeof(*eth));
      if (ip6h + 1 > data_end)
          return TC_ACT_UNSPEC;
     proto = ip6h->nexthdr;
      trans_data = ip6h + 1;
  }
   if (proto == IPPROTO_TCP)  {
       tcph = (struct tcphdr *)trans_data;
       dst = tcph->dst;
       src = tcph->source;
   } else if (proto == IPPROTO_UDP) {
       udph = (struct udphdr *)trans_data;
       dst = udph->dest;
       src = udph->source;
   } else {
     goto allow_unknown;
  }
   
```